src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AsyncLazy.cs

```diff
@@ -0,0 +1,33 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Roslyn.Utilities;
+
+internal static class AsyncLazy
+{
+    public static AsyncLazy<T, TArg> Create<T, TArg>(Func<TArg, CancellationToken, Task<T>> asynchronousComputeFunction, TArg arg)
+        => new AsyncLazy<T, TArg>(asynchronousComputeFunction, arg);
+
+    public static AsyncLazy<T, TArg> Create<T, TArg>(Func<TArg, CancellationToken, T> synchronousComputeFunction, TArg arg)
+        => new AsyncLazy<T, TArg>((arg, cancellationToken) => Task.FromResult(synchronousComputeFunction(arg, cancellationToken)), synchronousComputeFunction, arg);
```

Comment:
❓ If `AsyncLazy<T, TArg>` inherits from `AsyncLazy<T>`, why not do this?

```suggestion
    public static AsyncLazy<T> Create<T, TArg>(Func<TArg, CancellationToken, T> synchronousComputeFunction, TArg arg)
        => Create(
            asynchronousComputeFunction: static (outerArg, cancellationToken) => Task.FromResult(outerArg.synchronousComputeFunction(outerArg.arg, cancellationToken)), (synchronousComputeFunction, arg));
```